analyzers/src/SonarAnalyzer.CSharp/Rules/MemberInitializerRedundant.cs

```diff
@@ -63,168 +61,178 @@ public sealed class MemberInitializerRedundant : SonarDiagnosticAnalyzer
+                                return true;
+                            }
+
+                            return IsSymbolFirstSetInCfg(fieldOrPropertySymbol, constructor.Node, constructor.SemanticModel);
+                        }),
+                    DiagnosticMessage = InstanceMemberMessage
+                };
+
+            // Verifies the initializations done in static constructors and module initializers (C# 9).
+            public static readonly InitializationVerifier<BaseMethodDeclarationSyntax> StaticMemberInitializationVerifier =
+                new InitializationVerifier<BaseMethodDeclarationSyntax>
+                {
+                    InitializerFilter = method => method.MethodKind == MethodKind.StaticConstructor
+                                                  || method.IsModuleInitializer(),
+                    // only static members are interesting (const ones can only be initialized once)
+                    FieldFilter = field => field.Modifiers.Any(IsStatic),
+                    PropertyFilter = property => property.Modifiers.Any(IsStatic),
+                    IsSymbolFirstSetInInitializers = (fieldOrPropertySymbol, initializerDeclarations) =>
+                        // there can be only one static constructor
+                        // all module initializers are executed when the type is created, so it is enough if one initializes the member
+                        initializerDeclarations.Any(x => IsSymbolFirstSetInCfg(fieldOrPropertySymbol, x.Node, x.SemanticModel)),
+                    DiagnosticMessage = StaticMemberMessage
+                };
+
+            private static bool IsStaticOrConst(SyntaxToken token) =>
+                token.IsKind(SyntaxKind.StaticKeyword) || token.IsKind(SyntaxKind.ConstKeyword);
+
+            private static bool IsStatic(SyntaxToken token) => token.IsKind(SyntaxKind.StaticKeyword);
+
+            /// <summary>
+            /// Returns true if the member is overwritten without being read in the instance constructor.
+            /// Returns false if the member is not set in the constructor, or if it is read before being set.
+            /// </summary>
+            private static bool IsSymbolFirstSetInCfg(ISymbol classMember, BaseMethodDeclarationSyntax constructorOrInitializer, SemanticModel semanticModel)
+            {
+                var body = (CSharpSyntaxNode)constructorOrInitializer.Body ?? constructorOrInitializer.ExpressionBody();
+                if (!CSharpControlFlowGraph.TryGet(body, semanticModel, out var cfg))
+                {
+                    return false;
                 }
+
+                var checker = new MemberInitializerRedundancyChecker(cfg, classMember, semanticModel);
+                return checker.CheckAllPaths();
             }
         }
 
-        private static List<NodeSymbolAndSemanticModel<TSyntax, IMethodSymbol>> GetInitializerTuples<TSyntax>(SyntaxNodeAnalysisContext context, List<IMethodSymbol> constructorSymbols)
```

Comment:
[refactoring help] this has been moved to `InitializationVerifier` and renamed to `GetInitializerDeclarations`